app/src/lib/stores/app-store.ts

```diff
@@ -1646,50 +1648,85 @@ export class AppStore extends TypedBaseStore<IAppState> {
+   * `MissingRepository`, the menu labels will be updated but they will lack
+   * the expected `IRepositoryState` and revert to the default values.
    */
-  private updateMenuItemLabels(repository?: Repository) {
-    const editorLabel = this.selectedExternalEditor
-      ? `Open in ${this.selectedExternalEditor}`
-      : undefined
-
-    updatePreferredAppMenuItemLabels({
-      editorLabel: editorLabel,
-      pullRequestLabel: this.getPullRequestLabel(repository),
-      shellLabel: `Open in ${this.selectedShell}`,
-      defaultBranchName: this.getDefaultBranchName(repository),
-    })
-  }
+  private updateMenuLabelsForSelectedRepository() {
+    const { selectedState } = this.getState()
 
-  private getBranchesState(repository?: Repository) {
-    if (!repository || !repository.gitHubRepository) {
-      return undefined
+    if (
+      selectedState !== null &&
+      selectedState.type === SelectionType.Repository
+    ) {
+      this.updateMenuItemLabels(selectedState.state)
+    } else {
+      this.updateMenuItemLabels(null)
     }
-
-    const state = this.repositoryStateCache.get(repository)
-    return state.branchesState
   }
 
-  private getPullRequestLabel(repository?: Repository) {
-    const branchesState = this.getBranchesState(repository)
-    if (branchesState == null) {
-      return undefined
+  /**
+   * Update the menus in the main process using the provided repository state
+   *
+   * @param state the current repository state, or `null` if the repository is
+   *              being cloned or is missing
+   */
+  private updateMenuItemLabels(state: IRepositoryState | null) {
+    const shellLabel = `Open in ${this.selectedShell}`
+    const removeRepoLabel = this.getRemoveRepoLabel()
+
+    let labels: MenuLabels = {
+      shellLabel,
+      removeRepoLabel,
     }
 
-    if (branchesState.currentPullRequest === null) {
-      return undefined
+    if (this.selectedExternalEditor !== undefined) {
+      labels = {
+        ...labels,
+        editorLabel: `Open in ${this.selectedExternalEditor}`,
+      }
     }
 
-    return __DARWIN__ ? 'Show Pull Request' : 'Show &pull request'
+    if (state !== null) {
+      labels = {
+        ...labels,
+        pullRequestLabel: this.getPullRequestLabel(state),
+        defaultBranchName: this.getDefaultBranchName(state),
+        isForcePushForCurrentRepository: this.isCurrentBranchForcePush(state),
+        askForConfirmationOnForcePush: this.askForConfirmationOnForcePush,
+      }
+    }
+
+    updatePreferredAppMenuItemLabels(labels)
   }
 
-  private getDefaultBranchName(repository?: Repository) {
-    const branchesState = this.getBranchesState(repository)
-    if (branchesState == null) {
+  private isCurrentBranchForcePush(state: IRepositoryState) {
+    const { branchesState, aheadBehind } = state
+    return isCurrentBranchForcePush(branchesState, aheadBehind)
+  }
+
+  private getRemoveRepoLabel() {
+    return this.confirmRepoRemoval
```

Comment:
this is a bit nitpicky, but could we avoid the nested ternary here? (this is partly because it looks like `prettier` doesn't know how to format it properly, which makes it harder to read)

Reply:
Can we punt on this to a future PR? I want to stop passing strings from `AppStore` to the main process, and instead pass the information needed to decide what labels to show:

> * [bd69082](https://github.com/desktop/desktop/commit/bd690820efc6a7b2bc222c4bf9cec8e73badfaf8) - the first commit which passes through the original context, which we need to know what menu to show and what action to fire. I prefer this approach rather than maintaining two sets of string checks, and I'll come in later to follow up with other cleanup

Reply:
ah ok yes! missed that context.